File: 752-open-the-lock/752-open-the-lock.java
class Solution {
    public int openLock(String[] deadends, String target) {
        Set <String> deadset = new HashSet();
        Set <String> visited = new HashSet();
        int ans = 0;
        ArrayDeque<String> queue = new ArrayDeque();
        queue.addLast("0000");
        visited.add("0000");
        
        for (String deadend : deadends){
            deadset.add(deadend);
        }
        if (deadset.contains("0000")){
            return -1;
        }
        while (!queue.isEmpty()){
            int n = queue.size();
            ans++;
            
            for(int i = 0; i < n ; i++){
                StringBuilder current = new StringBuilder(queue.pollFirst());
                if (current.toString().equals(target)){
                    return ans - 1;
                }
                for (int k = 0 ; k < 4; k++){
                    char ch = current.charAt(k);
                    String build = current.substring(0, k) + ( ch == '9' ? 0 :  ch - '0' + 1)  + current.substring(k + 1);
                    if (!deadset.contains(build) && !visited.contains(build)){
                        queue.addLast(build);
                        visited.add(build);
                    }
                    build = current.substring(0, k) + ( ch == '0' ? 9 :  (ch - '0' - 1)) + current.substring(k + 1);
                    if (!deadset.contains(build) && !visited.contains(build)){
                        queue.addLast(build);
                        visited.add(build);
                    }
                }
            }
        }
        return -1;
    }
}

File: 752-open-the-lock/752-open-the-lock.py
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        if "0000" in deadends:
            return -1
        visited = set(deadends)
        begin = set()
        end = set()
        numSteps = 0
        begin.add("0000")
        end.add(target)
        
        while begin and end:
            temp = set()
            N = len(begin)
            for j in range(N):
                for move in list(begin):
                    if move in end:
                        return numSteps
                    if move in visited:
                        continue
                    visited.add(move)
                    for i in range(4):
                        curCharInc = move[:i] + str((int(move[i]) + 1)  % 10) + move[i+1:]
                        curCharDec = move[:i] + str((int(move[i]) - 1) % 10) + move[i+1:]
                        if curCharInc not in visited:
                            temp.add(curCharInc)
                        if curCharDec not in visited:
                            temp.add(curCharDec)
            numSteps += 1                
            begin = end.copy()
            end =  temp.copy()
        
        return -1